src/main/java/net/starlark/java/eval/RangeList.java

```diff
@@ -114,9 +114,24 @@ public StarlarkInt get(int index) {
 
   @Override
   public int size() {
+    Preconditions.checkState(size >= 0, "len(%s) = %s", this, len());
```

Comment:
Are we planning to patch `RangeList` to support full `StarlarkInt`? If so, `size` will either throw, or we should unimplement `Collection`.

Reply:
To support range over bigints would require len to return bigint, and I don't think the benefit is worth the cost if one can't create an actual array that big. After all, range loops are principally used to generate array indices (or to circumvent the recursion check).

Reply:
> range loops are principally used to generate array indices

`range` is also (infrequently) used with `in` operator:

```
Python 3.7.3 (default, Dec 13 2019, 19:58:14)
...
>>> (1 << 99) in range(1 << 88, 1 << 1000)
True
```

If you think that's not a reasonable scenario, that is OK, but it would make Starlark somewhat incomplete/inconsistent.

> I don't think the benefit is worth the cost

It can be implemented with nearly zero runtime cost:
* we have two `len` functions, one return `int`, other returns `StarlarkInt`. We can have different interfaces for different `len` functions
* `RangeList` may have two implementations: `RangeList.Int32` and `RangeList.Big`. First works with `int`, second works with `StarlarkInt`

The only cost is the implementation/maintenance cost.

Reply:
> The only cost is the implementation/maintenance cost.

That's the cost I'm concerned about. We already make concessions to the representation choices of the host language in the specification of strings; I think it's fine to do it for array lengths too. Object[].length and Collections.size return int. Even Python imposes limits due to the widths of host data types; for example:

```list(range(1<<63)) # OverflowError: Python int too large to convert to C ssize_t```

Reply:
> list(range(1<<63)) # OverflowError: Python int too large to convert to C ssize_t

you probably meant `len`, not `list`

```
>>> len(range(1<<63))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: Python int too large to convert to C ssize_t
```

which is TIL, and I expected better from Python.

So, TL;DR, are you suggesting:
* keeping len <= maxint
* failing at construction if len exceeds maxint
* allowing or prohibiting ranges with len <= maxint, when start > maxint?

Reply:
I did mean len, thanks (though apparently list fails with the same error).

Yes, I suggest we disallow construction of a range whose len exceeds maxint. We could easily support "short" ranges of arbitrarily large ints, but that could be an independent change, and it doesn't seem urgent.